### PR TITLE
 DL: Update content of the model summary table

### PR DESCRIPTION
**Pull request by njayaram2**
JIRA: MADLIB-1349, MADLIB-1338
Add a new column named `metrics_type` that captures the metric name
(if) specified in the compile params. This commit also fixes the
model_size reported, it was reporting the wrong size.

There were a few comments in MADLIB-1338 after the corresponding PR was
merged to master. Some of those comments are addressed in this PR since
they are related code, and minor changes.
1. Ensure compile_params cannot be NULL.
2. Update some info messages.
3. Use the @MinWarning decorator to suppress notices.
4. Ensure training_metrics_final and validation_metrics_final are NULL
if the compile params did not have any metric specified.

Co-authored-by: Jingyi Mei <jmei@pivotal.io>

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/932/

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 1]
```diff
@@ -238,19 +223,22 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
+    metrics_list = get_metrics_from_compile_param(compile_params)
+    is_metrics_specified = True if metrics_list else False
+    metrics_type = 'ARRAY{0}'.format(metrics_list) if is_metrics_specified else 'NULL'
     metrics_iters = metrics_iters if metrics_iters else 'NULL'
     # We always compute the training loss and metrics, at least once.
-    training_metrics_final = training_metrics[-1]
     training_loss_final = training_loss[-1]
-    training_metrics = training_metrics if training_metrics else 'NULL'
-    training_loss = training_loss if training_loss else 'NULL'
+    training_loss = 'ARRAY{0}'.format(training_loss) if training_loss else 'NULL'
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: can't we  avoid this if check if we use plpy prepare for training_loss and others ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 2]
```diff
@@ -319,6 +308,38 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
+                {class_values} AS class_values,
+                {norm_const} AS norm_const,
+                {dep_vartype} AS dep_vartype,
+                {dep_varname} AS dependent_varname_in_source_table,
+                {indep_varname} AS independent_varname_in_source_table
+            FROM {tbl}
+        """.format(class_values=CLASS_VALUES_COLNAME,
+                   norm_const=NORMALIZING_CONST_COLNAME,
+                   dep_vartype=DEPENDENT_VARTYPE_COLNAME,
+                   dep_varname='dependent_varname',
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: don't we have variables for indep and dep varname constants ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 3]
```diff
@@ -85,14 +66,17 @@ def fit(schema_madlib, source_table, model,model_arch_table,
     independent_varname = MINIBATCH_OUTPUT_INDEPENDENT_COLNAME_DL
     model_arch_table = quote_ident(model_arch_table)
     fit_params = "" if not fit_params else fit_params
+    compile_params = "" if not compile_params else compile_params
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: This does not, but compile_params is quoted using `$madlib$` before passing it along to the transition function. We cannot do that to `None`, so this only changes None to an empty string. The actual validation happens in `parse_and_validate_compile_params()` in `madlib_keras_wrapper.py_in`.

In the current state of the code, compile params are validated only in `fit_transition`. That's not ideal, but I think there is some other JIRA to refactor that code and ensure validation happens in `fit`.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 4]
```diff
@@ -85,14 +66,17 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
         source_table, validation_table, model, model_arch_table,
-        dependent_varname, independent_varname, num_iterations,
-        metrics_compute_frequency)
+        dependent_varname, independent_varname,
+        num_iterations, metrics_compute_frequency)
     if metrics_compute_frequency is None:
         metrics_compute_frequency = num_iterations
 
+    # The following two times must be recorded together.
+    metrics_elapsed_start_time = time.time()
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: The variable names reflect the corresponding column names in the output summary table, so I would actually prefer keeping `start_training_time` as is.

File: src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in [hunk 5]
```diff
@@ -160,7 +172,7 @@ def parse_and_validate_compile_params(str_of_args):
         opt_args:               Arguments for the optimizer
         compile_dict:           Dictionary of arguments for keras.compile
     """
-
+    _assert(str_of_args, "Compile parameters cannot be empty or NULL.")
```
[the remark this answers is not in the document]
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras_wrapper.py_in]: True, the value here will not be NULL, but the value passed by the user can be either empty or NULL. Since we convert a `None` to empty string in the other code you pointed out, we won't know the original value exactly here. So made the error message generic.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 1 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: That's correct, but re-ordering the columns becomes a pain if we populate a lot of values using prepare. We have over 20 columns in the summary table and 6 of them are already populated using prepare.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 2 above)
Comment (njayaram2) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: I think the `dependent_varname` variable is equal to `MINIBATCH_OUTPUT_DEPENDENT_COLNAME_DL` (the column name of the dependent variable in the minibatched table), whereas the string `'dependent_varname'` here refers to the column name in the minibatched summary table that stores the value of the dependent variable column name (say `y`) in the original non-minibatched table.
Confusing, but this is another refactor JIRA by itself I guess!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/940/

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: We can do the null compile param check here instead of line 175 in `madlib_keras_wrapper.py_in `

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/942/
